### PR TITLE
all core features

### PR DESCRIPTION
**Pull request by amohamed97**

### CODE REVIEW

File: ui/src/components/grader/DownloadResult.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import React from 'react'
+import axios from 'axios'
+import fileDownload from 'js-file-download'
+
+
+function DownloadResult(props){
+
+    function downloadFile(){
+        axios.get(`http://localhost:5000/results/cc451/${props.lab}`).then(res => {
```
Comment (shakram02) [inline ui/src/components/grader/DownloadResult.js]: please extract the website URL to a `config` file or something. Where it tries to fetch the variable from an environment variable first, then falls back to a default value

File: ui/src/components/grader/DownloadResult.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    return (
+        <React.Fragment>
+            <hr></hr>
+            <p>Results are ready for download</p>
+            <button onClick={downloadFile}>DOWNLOAD</button>
+        </React.Fragment>
+    )
+}
+
+export default DownloadResult
```
Comment (shakram02) [inline ui/src/components/grader/DownloadResult.js]: please make sure there's an empty line at the end of each file, a formatter extension will do that

File: ui/src/components/grader/FileUpload.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    function fileData(){ 
+        if (selectedFile) { 
+            return ( 
+                <div> 
+                    <h2>File Details:</h2> 
+                    <p>File Name: {selectedFile.name}</p> 
+                    <p>File Type: {selectedFile.type}</p> 
+                    <p> 
+                    Last Modified:{" "}
```
Comment (shakram02) [inline ui/src/components/grader/FileUpload.js]: We don't really need the `lastModified` date

File: ui/src/components/grader/FileUpload.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        selectFile(event.target.files[0])
+    }; 
+
+    function onFileUpload(){ 
+        const formData = new FormData(); 
+        formData.append( 
+            "submissions_file", 
+            selectedFile
+        ); 
+        axios.post(`http://localhost:5000/results/cc451/${props.lab}`, formData).then((res) =>{
```
Comment (shakram02) [inline ui/src/components/grader/FileUpload.js]: please extract the main URL

File: ui/src/components/grader/LabSelector.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+
+function LabSelector(props) {
+    const classes = useStyles();
+    const [currLab, setCurrLab] = useState('');
+    const [labs, setLabs] = useState([]);
+
+
+    useEffect(() => {
+        Axios.get(`http://localhost:5000/labs/cc451`).then((res) => {
```
Comment (shakram02) [inline ui/src/components/grader/LabSelector.js]: extract (I'll repeat the comment so you can resolve it once you're done)

File: ui/src/components/grader/Grader.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                        }
+                        
+                    </Grid>
+                </Paper>
+            </Grid>
+        </Grid>
+    )
+}
+
+export default Grader
```
Comment (shakram02) [inline ui/src/components/grader/Grader.js]: new line (I'll repeat the comment so you can resolve it once you're done)

File: ui/src/components/grader/LabSelector.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+                    label="Select Lab"
+                >
+                    {labs.map(lab => <MenuItem key={lab} value={lab}>{lab}</MenuItem>)}
+                </Select>
+            </FormControl>
+        </React.Fragment>
+    )
+}
+
+export default LabSelector
```
Comment (shakram02) [inline ui/src/components/grader/LabSelector.js]: new line

File: ui/src/components/layout/Header.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+            <Toolbar>
+                <Typography variant="h6" color="inherit" noWrap>
+                    GraderX
+                </Typography>
+            </Toolbar>
+        </AppBar>
+    )
+}
+
+export default Header
```
Comment (shakram02) [inline ui/src/components/layout/Header.js]: new line

File: ui/src/components/grader/Grader.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+    function changeLab(val){
+        setLab(val)
+        setResultsReady(false)
+    }
+
+    return (
+        <Grid container justify="center" alignItems="center">
+            <Grid item>
+                <Paper elevation={5} style={{ 'padding': 25, 'minWidth':700}}>
```
Comment (shakram02) [inline ui/src/components/grader/Grader.js]: will this work fine on a mobile screen? `minWidth:700`

File: ui/src/components/grader/FileUpload.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+                }
+				
+			</div> 
+		    {fileData()} 
+		</div> 
+	); 
+
+}
+
+export default FileUpload;
```
Comment (shakram02) [inline ui/src/components/grader/FileUpload.js]: new line

**Review by shakram02 — COMMENTED** [review]
Great job :clap:

File: src/__tests__/__snapshots__/FileUpload.js.snap [hunk 11]
```diff
@@ -0,0 +1,21 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (shakram02) [inline src/__tests__/__snapshots__/FileUpload.js.snap]: do we need this file?

File: src/__tests__/FileUpload.js [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+import FileUpload from '../components/grader/FileUpload';
+import renderer from 'react-test-renderer';
+
+
+test('Dummy Test', () => {
+  const component = renderer.create(<FileUpload/>);
+  let tree = component.toJSON();
+  expect(tree).toMatchSnapshot();
+  
+});
```
Comment (shakram02) [inline src/__tests__/FileUpload.js]: add new line :3

File: src/components/grader/FileUpload.js [hunk 13]
```diff
@@ -9,21 +9,41 @@ function FileUpload(props) {
...
+      alert("Successfully Uploaded...")
+      return false;
+    }
+    
   }
 
+  function validateExtention(){
+
+    let extension = selectedFile.type
+    let supportedTypes = ["application/zip", "application/vnd.rar", "application/x-7z-compressed" ];
```
Comment (shakram02) [inline src/components/grader/FileUpload.js]: extract this to a const array, since it's never changing.

**Review by shakram02 — CHANGES_REQUESTED** [review]

File: src/__tests__/__snapshots__/FileUpload.js.snap (hunk 11 above)
Comment (Hagar-Usama) [reply src/__tests__/__snapshots__/FileUpload.js.snap]: Nope, removed + added to .gitignore

**Comment by shakram02** [comment]
why is `add upload spinner` in this branch?, please remove it for this branch to be merged

**Pull request merged**